src/lib/propagation/DAG.js

```diff
@@ -0,0 +1,63 @@
+import * as NodeFns from './nodeFns'
+import {SimulationNode} from './node'
+import * as constants from './constants'
+
+import * as _collections from 'gEngine/collections'
+
+export class SimulationDAG {
+  constructor(nodes) {
+    if (!!_.get(window, 'recorder')) { window.recorder.recordSimulationDAGConstructionStart(this) }
+    let rest = nodes.map(NodeFns.extractInputs)
+
+    let byId = _.transform(rest, (map, node) => {map[node.id] = {node, lastAncestors: node.inputs, ancestors: node.inputs}}, {})
+
+    const missingInputsNodes = _.remove(rest, NodeFns.hasMissingInputs(rest))
+    let graphErrorNodes = missingInputsNodes.map(NodeFns.withMissingInputError(nodes))
```

Comment:
How so? Happy to make less opaque, but I'm not exactly sure where you're confused. What they do is first extract all nodes from the list that have missing inputs (i.e. inputs not contained in the passed in list of nodes), then add those to the list of graph error nodes.